src/bp/core/services/messaging/conversations.ts

```diff
@@ -0,0 +1,120 @@
+    @inject(TYPES.ConversationRepository) private conversationRepo: ConversationRepository
+  ) {}
+
+  @postConstruct()
+  async init() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.CONFIGURATION_LOADED)
+
+    this.invalidateMostRecent = <any>await this.jobService.broadcast<void>(this._localInvalidateMostRecent.bind(this))
+  }
+
+  private _localInvalidateMostRecent(botId: string, userId: string, mostRecentConvoId?: sdk.uuid) {
+    this.forBot(botId).localInvalidateMostRecent(userId, mostRecentConvoId)
+  }
+
+  public forBot(botId: string): ScopedConversationService {
+    let scope = this.scopes[botId]
+    if (!scope) {
+      scope = new ScopedConversationService(botId, this.conversationRepo, (userId, mostRecentConvoId) =>
+        this.invalidateMostRecent(botId, userId, mostRecentConvoId)
+      )
+      this.scopes[botId] = scope
+    }
+    return scope
+  }
+}
+
+export class ScopedConversationService implements sdk.experimental.conversations.BotConversations {
+  private mostRecentCache: LRU<string, sdk.experimental.Conversation>
+
+  constructor(
+    private botId: string,
+    private conversationRepo: ConversationRepository,
+    public invalidateMostRecent: (userId: string, mostRecentConvoId?: sdk.uuid) => void
+  ) {
+    this.mostRecentCache = new LRU<string, sdk.experimental.Conversation>({ max: 10000, maxAge: ms('5min') })
+  }
+
+  public async list(
+    filters: sdk.experimental.conversations.ListFilters
+  ): Promise<sdk.experimental.RecentConversation[]> {
+    return this.conversationRepo.list(this.botId, filters)
+  }
+
+  public async delete(filters: sdk.experimental.conversations.DeleteFilters): Promise<number> {
+    if (filters.id) {
+      const conversation = (await this.conversationRepo.get(filters.id))!
+      this.invalidateMostRecent(conversation.userId)
+
+      return (await this.conversationRepo.delete(filters.id)) ? 1 : 0
+    } else {
+      this.invalidateMostRecent(filters.userId!)
+
+      return this.conversationRepo.deleteAll(this.botId, filters.userId!)
+    }
+  }
+
+  public async create(args: sdk.experimental.conversations.CreateArgs): Promise<sdk.experimental.Conversation> {
+    return this.conversationRepo.create(this.botId, args)
+  }
+
+  public async recent(
+    filters: sdk.experimental.conversations.RecentFilters
+  ): Promise<sdk.experimental.RecentConversation> {
+    const { userId } = filters
+    const cached = this.mostRecentCache.get(userId)
+    if (cached) {
+      return cached
+    }
+
+    let conversation = await this.conversationRepo.recent(this.botId, filters)
+    if (!conversation) {
+      conversation = await this.conversationRepo.create(this.botId, { userId })
+    }
+
+    this.mostRecentCache.set(userId, conversation)
+
+    return conversation
+  }
+
+  public async get(
+    filters: sdk.experimental.conversations.GetFilters
+  ): Promise<sdk.experimental.Conversation | undefined> {
+    return this.conversationRepo.get(filters.id)
+  }
```

Comment:
Same, as parameters for all of these ^

Reply:
Yup I removed filters for get it was a bit overkill. Also for recent